website/blog/2022-10-31-simple-1.mdx

```diff
@@ -0,0 +1,102 @@
+
+#### Null setting avoidance
+The adapter can statically infer if a function never generates null; In the simple function interface if the `call` function return's type is `void`, it means the output is never null, and if it's `bool`, then the function returns true for not null and false for null).
+When the function does not generate nulls, then null setting is **completely avoided** during the computation (only the previous bull setting is needed). The consequence of that is that the hot loop applying the function becomes simdizable triggering a huge boost in performance for primitive operations.
+
+
+#### Encoding based fast path
+Vectors in Velox can have different encodings (flat, constant..etc). The generic way of reading a vector of arbitrary encoding is to use a decoded vector to guarantee correct data access. That translates into an overhead each time an input value is accessed (we need to check the encoding of the vector to know how to read the value for every row).
+
+When the function is a primitive operation like plus or minus, such overhead is expensive! To avoid that, encoding based fast paths can be added, the code snippet below illustrates the idea.
+<figure>
+<img src="/img/simple1_11.png" width="60%"/>
+</figure>
+
+In the code above, the overhead of checking the encoding is switched outside the loop that applies the functions (the plus operation here). And the inner loops are simple operations that are potentially simdizable and free of encoding checks.
+One issue with this optimization is that the core loop is replicated many times. In general, the numbers of times it will be replicated
+is `n^m` where `n` is the number of args, and `m` is the number of encodings.
+
+To avoid code size blowing, we only apply this optimization when all input arguments are primitives and the number of input arguments is <=3.
+The figure below shows the effect of this optimization on the processing time of a query of primitive operations.
+<figure>
+<img src="/img/simple1_12.png"  width="70%"  />
+</figure>
+To compromise for both (performance and code size) when the conditions for specializing for all encodings are not met, we have a pseudo specialization mode that does not blow up the code size, but still reduce the overhead of decoding to a single multiplication per argument. This mode is enabled when all the primitive arguments are either flat or constant. The code below illustrates the idea:
+<figure>
+<img src="/img/simple1_13.png" width="60%" />
+</figure>
+
+When the input vector is constant we can read the value always from index 0 of the values buffer, and when it is flat  we can read it from the index row; this can be achieved by assigning a factor to either 0 or 1 and reducing the decoding operation per row into a multiplication with that factor Note that such a multiplication does not prevent simd.
```

Comment:
not required, but would be awesome if we could provide some number on how much faster these optimizations make the system. Do we have micro-benchmarks that test these behaviors?

Reply:
already included in the graph